runtime/dotnet/core/SkillConversationIdFactory.cs

```diff
@@ -1,45 +1,78 @@
 // Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License.
 
-using System.Collections.Concurrent;
+using System;
+using System.Collections.Generic;
+using System.Linq;
 using System.Threading;
 using System.Threading.Tasks;
+using Microsoft.Bot.Builder;
 using Microsoft.Bot.Builder.Skills;
-using Microsoft.Bot.Schema;
-using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
 
 namespace Microsoft.BotFramework.Composer.Core
 {
     /// <summary>
-    /// A <see cref="SkillConversationIdFactory"/> that uses an in memory <see cref="ConcurrentDictionary{TKey,TValue}"/>
-    /// to store and retrieve <see cref="ConversationReference"/> instances.
+    /// A <see cref="SkillConversationIdFactory"/> that uses <see cref="IStorage"/> to store
+    /// and retrieve <see cref="SkillConversationReference"/> instances.
     /// </summary>
     public class SkillConversationIdFactory : SkillConversationIdFactoryBase
     {
-        private readonly ConcurrentDictionary<string, string> _conversationRefs = new ConcurrentDictionary<string, string>();
+        private readonly IStorage _storage;
 
-        public override Task<string> CreateSkillConversationIdAsync(SkillConversationIdFactoryOptions options, CancellationToken cancellationToken)
+        public SkillConversationIdFactory(IStorage storage)
         {
+            _storage = storage ?? throw new ArgumentNullException(nameof(storage));
+        }
+
+        public override async Task<string> CreateSkillConversationIdAsync(SkillConversationIdFactoryOptions options, CancellationToken cancellationToken)
+        {
+            if (options == null)
+            {
+                throw new ArgumentNullException(nameof(options));
+            }
+
+            // Create the storage key based on the SkillConversationIdFactoryOptions.
+            var conversationReference = options.Activity.GetConversationReference();
+            var skillConversationId = $"{options.FromBotId}-{options.BotFrameworkSkill.AppId}-{conversationReference.Conversation.Id}-{conversationReference.ChannelId}-skillconvo";
```

Comment:
what does this "skillconvo" mean?

Reply:
It is just an arbitrary string, so you know that is a skill conversation ID when debugging, it not required at all, is up to the dev what the ID looks like.